chakrashim: build on Linux

jianchun opened the pull request:
This initial change enables building node-chakracore on Linux, based on latest cross-platform effort of ChakraCore.

To build:

```
./configure --engine=chakracore
make
```

To run:

```
./node
```

Primary commit is https://github.com/nodejs/node-chakracore/commit/c3f8b83

kunalspathak commented on deps/chakrashim/chakrashim.gyp:
```diff
@@ -45,14 +48,18 @@
           [ 'target_arch=="arm"', {
             'defines': [ '__arm__=1' ]
           }],
+          ['OS == "win"', {
```
curious, what is this for?

jianchun replied:
Only on Windows link to ole32.lib/version.lib

kunalspathak commented on deps/chakrashim/src/jsrtisolateshim.cc:
```diff
@@ -224,8 +224,7 @@ void IsolateShim::PushScope(
   // Don't crash even if we fail to set the context
-  JsErrorCode errorCode = JsSetCurrentContext(contextShim->GetContextRef());
-  CHAKRA_ASSERT(errorCode == JsNoError);
+  JsSetCurrentContext(contextShim->GetContextRef());
```
do we need similar thing (ignore errorCode) in chakracore-master as well?

jianchun replied:
The error code was already ignored, just a debug assert. I did this because compiler gives warnings variable "errorCode" is not used. Alternatively could use some macro "UNUSED(errorCode)" to avoid those warnings but didn't find an existing macro, and thought not very useful for an assertion plus another macro and they don't do much.

kunalspathak replied:
For debug build they are not ignored given that they are used in assert to make sure the value is `JsNoError`. We could have something like this

``` c++
#ifdef DEBUG
JsErrorCode errorCode = 
#endif
JsSetCurrentContext(contextShim->GetContextRef());
CHAKRA_ASSERT(errorCode == JsNoError); // This line will be removed for release.
```

jianchun replied:
That'll work too. I can revert all these CHAKRA_ASSERTs if you feel they are useful.

kunalspathak replied:
Yeah, it will be good to have them around to find bugs.

kunalspathak commented on deps/chakrashim/include/v8.h:
```diff
@@ -1585,21 +1583,21 @@ class FunctionCallbackInfo {
   }
 
  private:
+  Value** _args;
```
why have we moved fields here?

jianchun replied:
Default compiler complains if fields declared in the class header and field initialization lists in constructor have different order.

kunalspathak replied:
i see and `msbuild` doesn't complain on windows. Interesting.

jianchun replied:
There must be a compiler setting to disable these warnings. Didn't bother to tweak that; re-ordering is easy.

kunalspathak commented on deps/chakrashim/src/jsrtstringutils.cc:
```diff
@@ -23,175 +23,175 @@
-  //
-  return JsNoError;
-}
+// JsErrorCode StringConvert::GetCharLength(const wchar_t *str,
```
add comment on why this is commented out.

jianchun replied:
These are described in the commit message and need new solutions for xplat. They use Windows only APIs. I could "#ifdef" them for Win32. Just quickly commented them all to be able to build first.

kunalspathak replied:
Yes, I saw your commit message, but i was saying it would make more sense to add that comment in the file as well or have `#ifdef` for win32.

kunalspathak commented on deps/chakrashim/src/v8function.cc:
```diff
@@ -87,7 +87,7 @@ Local<Value> Function::Call(Handle<Value> recv,
 
 void Function::SetName(Handle<String> name) {
-  JsErrorCode error = DefineProperty((JsValueRef)this,
+  DefineProperty((JsValueRef)this,
```
why ignore error?

jianchun replied:
Same as before, "error" was only referenced in a debug assert and not used in Release build, compiler complains.

kunalspathak commented on deps/chakrashim/src/v8handlescope.cc:
```diff
@@ -42,8 +42,7 @@ HandleScope::~HandleScope() {
     // Don't crash even if we fail to release the scope
-    JsErrorCode errorCode = JsRelease(currRecord->_ref, nullptr);
-    CHAKRA_ASSERT(errorCode == JsNoError);
+    JsRelease(currRecord->_ref, nullptr);
```
same here, should this be patched in chakracore-master as well?

jianchun replied:
No need to. This only removes a compiler warning. Original code help debugging a little bit. Alternatively can change them all to "CHAKRA_ASSERT(...); UNUSED(errorCode);". Just felt the assertions are not that important in most of the cases.

kunalspathak commented on deps/chakrashim/src/v8handlescope.cc:
```diff
@@ -23,13 +23,13 @@
 HandleScope::HandleScope(Isolate* isolate)
-    : _prev(current),
-      _locals(),
+    : _locals(),
```
why moved this around?

jianchun replied:
Same as other cases, to align declaration and initialization order to make compiler happy.

Commit:
chakrashim: patch to build with linux chakracore

Update build files to build with linux ChakraCore.

PR-URL: https://github.com/nodejs/node-chakracore/pull/83
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

Commit:
chakrashim: build on Linux

Initial changes to build node-chakracore on Linux.

```
./configure --engine=chakracore
make

./node
```

Primary changes:
- Hook gyp with `build.sh` to build ChakraCore. Currently hard coded to
  Release build only. Link with static ChakraCore libs on xplat.

- Fix all chakrashim code to build on Linux (compiler differences).

- Change chakrashim to use new JSRT Utf8 APIs.

NOTEs:
- Original `Utf8/OneByte/TwoByte` encoding conversion code depends
  on Win32 APIs that are unavailable xplat. This change removes them and
  regresses related functionalities. Will need future work to improve.

Unit test results:

Before change, Windows: [+1165 -21]
After change, Windows:  [+1148 -38]
After change, Linux:    [+1113 -47]

PR-URL: https://github.com/nodejs/node-chakracore/pull/83
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

Commit:
chakrashim: remove --engine test arg

Detect engine from tests automatically instead.

PR-URL: https://github.com/nodejs/node-chakracore/pull/83
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

Commit:
chakrashim: switch to chakracore/linux

For xplat, switch deps/chakrashim/core to latest ChakraCore linux branch
on 7/12/2016 at https://github.com/Microsoft/ChakraCore/commit/139c50d

PR-URL: https://github.com/nodejs/node-chakracore/pull/83
Reviewed-By: Kunal Pathak <Kunal.Pathak@microsoft.com>

kunalspathak commented on tools/test.py:
```diff
@@ -875,6 +874,11 @@ def GetVmFlags(self, testcase, mode):
   def GetTimeout(self, mode):
     return self.timeout * TIMEOUT_SCALEFACTOR[ARCH_GUESS or 'ia32'][mode]
 
+  def GetEngine(self, arch, mode):
```
may be we could port this to chakracore-master as well.

jianchun replied:
> may be we could port this to chakracore-master as well.

Yes. (The reason I changed this is because we don't have an external passed in "engine = ..." switch.)
